lib/ecto/repo/schema.ex

```diff
@@ -706,7 +706,7 @@ defmodule Ecto.Repo.Schema do
   defp dump_field!(action, schema, field, type, value, adapter) do
     case Ecto.Type.adapter_dump(adapter, type, value) do
       {:ok, value} ->
-        {field, value}
+        {schema.__schema__(:source, field), value}
```

Comment:
@michalmuskala - If i understand it correctly, the current types are stored in the changeset. So, we should also add a key `source` in the changeset similar to `types` and read the source from there.